File: app/services/conversation/processed_update_flow.rb
class Conversation::ProcessedUpdateFlow
  def initialize(conversation_service:)
    @conversation_service = conversation_service
  end

  def find(update_id)
    @conversation_service.get_processed_update(update_id)
  end

  def begin_processing(message)
    claimed_update_ids = []

    message.processing_updates.each do |processing_update|
      claimed = @conversation_service.begin_processing(
        processing_update.fetch(:update_id),
        message.chat_id,
        processing_update.fetch(:message_id)
      )
      return rollback_processing_claims(claimed_update_ids) unless claimed

      claimed_update_ids << processing_update.fetch(:update_id)
    end

    true
  end

  def clear_processing(message)
    message.processing_updates.each do |processing_update|
      @conversation_service.clear_processing(processing_update.fetch(:update_id))
    end
  end

  def duplicate?(processed_update)
    processed_update&.sent_at.present?
  end

  def replayable?(processed_update)
    processed_update&.reply_text.present? && processed_update.conversation_state.present?
  end

  def resend_pending_reply(message, processed_update, telegram_client:)
    telegram_client.send_message(
      message.chat_id,
      processed_update.reply_text,
      suggested_replies: parse_suggested_replies(processed_update.suggested_replies)
    )
    @conversation_service.persist_conversation_state(message.chat_id, processed_update.conversation_state)
    mark_processed(message)
  end

  def mark_processed(message)
    message.processing_updates.each do |processing_update|
      @conversation_service.mark_processed(
        processing_update.fetch(:update_id),
        message.chat_id,
        processing_update.fetch(:message_id)
      )
    end
  end

  private

  def rollback_processing_claims(claimed_update_ids)
    claimed_update_ids.each do |update_id|
      @conversation_service.clear_processing(update_id)
    end

    false
  end

  def parse_suggested_replies(raw_suggested_replies)
    return [] if raw_suggested_replies.blank?

    parsed_replies = JSON.parse(raw_suggested_replies)
    return [] unless parsed_replies.is_a?(Array)

    parsed_replies.filter_map do |reply|
      next unless reply.is_a?(String)

      normalized_reply = reply.strip
      next if normalized_reply.empty?

      normalized_reply
    end
  rescue JSON::ParserError
    []
  end
end

File: src/main/java/com/telegramcodex/conversation/ProcessedUpdateFlow.java
package com.telegramcodex.conversation;

import com.telegramcodex.codex.ReplyParser;
import com.telegramcodex.telegram.InboundMessage;
import com.telegramcodex.telegram.TelegramClient;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import org.springframework.stereotype.Component;

@Component
public class ProcessedUpdateFlow {

    private final ConversationService conversationService;
    private final ReplyParser replyParser;

    public ProcessedUpdateFlow(ConversationService conversationService, ReplyParser replyParser) {
        this.conversationService = conversationService;
        this.replyParser = replyParser;
    }

    public Optional<ProcessedUpdateRecord> find(long updateId) {
        return conversationService.getProcessedUpdate(updateId);
    }

    public boolean beginProcessing(InboundMessage message) {
        ArrayList<Long> claimedUpdateIds = new ArrayList<>();
        for (InboundMessage.ProcessingUpdate processingUpdate : message.processingUpdates()) {
            boolean claimed = conversationService.beginProcessing(
                processingUpdate.updateId(),
                message.chatId(),
                processingUpdate.messageId()
            );
            if (!claimed) {
                rollbackProcessingClaims(claimedUpdateIds);
                return false;
            }
            claimedUpdateIds.add(processingUpdate.updateId());
        }
        return true;
    }

    public boolean duplicate(Optional<ProcessedUpdateRecord> processedUpdate) {
        return processedUpdate.map(ProcessedUpdateRecord::sentAt).orElse(null) != null;
    }

    public boolean replayable(Optional<ProcessedUpdateRecord> processedUpdate) {
        return processedUpdate
            .filter(update -> update.replyText() != null && update.conversationState() != null)
            .isPresent();
    }

    public void resendPendingReply(InboundMessage message, ProcessedUpdateRecord processedUpdate, TelegramClient telegramClient) {
        telegramClient.sendMessage(
            message.chatId(),
            processedUpdate.replyText(),
            replyParser.parseSuggestedReplies(processedUpdate.suggestedReplies()),
            false
        );
        conversationService.persistConversationState(message.chatId(), processedUpdate.conversationState());
        markProcessed(message);
    }

    public void markProcessed(InboundMessage message) {
        for (InboundMessage.ProcessingUpdate processingUpdate : message.processingUpdates()) {
            conversationService.markProcessed(processingUpdate.updateId(), message.chatId(), processingUpdate.messageId());
        }
    }

    private void rollbackProcessingClaims(List<Long> claimedUpdateIds) {
        for (Long updateId : claimedUpdateIds) {
            conversationService.clearProcessing(updateId);
        }
    }
}
